crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
+      }
+
+  pkey->pkey.nistdsa_key = key;
+  return 1;
+}
+
+static int pkey_nistdsa_sign_message(EVP_PKEY_CTX *ctx, uint8_t *sig,
+                                        size_t *siglen, const uint8_t *tbs,
+                                        size_t tbslen) {
+  NISTDSA_PKEY_CTX *dctx = ctx->data;
+  const NISTDSA *nistdsa = dctx->nistdsa;
+  if (nistdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    nistdsa = NISTDSA_KEY_get0_sig(ctx->pkey->pkey.nistdsa_key);
   }
+
   // Caller is getting parameter values.
   if (sig == NULL) {
-    *siglen = DILITHIUM3_SIGNATURE_BYTES;
+    *siglen = nistdsa->signature_len;
     return 1;
   }
 
-  if (*siglen < DILITHIUM3_SIGNATURE_BYTES) {
+  if (*siglen < nistdsa->signature_len) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
     return 0;
   }
 
-  if (ml_dsa_65_sign(sig, siglen, tbs, tbslen, NULL, 0, key->priv) != 0) {
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+      ctx->pkey->pkey.nistdsa_key == NULL ||
+      ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+      }
+
+  NISTDSA_KEY *key = ctx->pkey->pkey.nistdsa_key;
+  if (!key->secret_key) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+
+  if (nistdsa->method->sign(sig, siglen, tbs, tbslen, NULL,
+                            0, key->secret_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, ERR_R_INTERNAL_ERROR);
     return 0;
   }
   return 1;
 }
 
-static int pkey_dilithium3_verify_message(EVP_PKEY_CTX *ctx, const uint8_t *sig,
+static int pkey_nistdsa_verify_message(EVP_PKEY_CTX *ctx, const uint8_t *sig,
                                           size_t siglen, const uint8_t *tbs,
                                           size_t tbslen) {
-  if (ctx == NULL || ctx->pkey->pkey.ptr == NULL) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_MISSING_PARAMETERS);
-    return 0;
+  NISTDSA_PKEY_CTX *dctx = ctx->data;
+  const NISTDSA *nistdsa = dctx->nistdsa;
+
+  if (nistdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+
+    nistdsa = NISTDSA_KEY_get0_sig(ctx->pkey->pkey.nistdsa_key);
   }
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+    ctx->pkey->pkey.nistdsa_key == NULL ||
+    ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+    }
 
-  DILITHIUM3_KEY *key = ctx->pkey->pkey.ptr;
+  NISTDSA_KEY *key = ctx->pkey->pkey.nistdsa_key;
 
-  if (siglen != DILITHIUM3_SIGNATURE_BYTES ||
-      ml_dsa_65_verify(tbs, tbslen, sig, siglen, NULL, 0, key->pub) != 0) {
+  if (siglen != nistdsa->signature_len ||
+      nistdsa->method->verify(tbs, tbslen, sig, siglen,
+                              NULL, 0, key->public_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_SIGNATURE);
```

Comment:
I was following the same expected behaviour in https://github.com/jakemas/aws-lc/blob/3f03f7d7b2eccca69f278feaeb073ab9a5540c61/crypto/fipsmodule/evp/p_ed25519.c#L77
```
  if (siglen != 64 ||
      !ED25519_verify(tbs, tbslen, sig, key->key + ED25519_PUBLIC_KEY_OFFSET)) {
    OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_SIGNATURE);
    return 0;
  }
```
How would you like this to be handled?

Reply:
I checked what happens with ECDSA and it's even worse, it can return both `ECDSA_R_BAD_SIGNATURE` and `ECDSA_R_MISMATCHED_SIGNATURE` but no `EVP_R_INVALID_SIGNATURE `.

Reply:
I guess let's stay consistent and keep `EVP_R_INVALID_SIGNATURE`.

Reply:
ok